### PR TITLE
[BEAM-4176] Validate Runner Tests generalization and enable for local reference runner

### PR DESCRIPTION
**Pull request by angoenka**
**Please** add a meaningful description for your change here

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | --- | ---

### CODE REVIEW

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/job/ReferenceRunnerJobService.java [hunk 1]
```diff
@@ -200,5 +218,17 @@ public void close() throws Exception {
...
+          artifactStagingServices.remove();
+      try {
+        artifactStagingService.close();
+      } catch (Exception e) {
+        LOG.error(
+            "Unable to close staging sevice started on %s",
+            artifactStagingService.getApiServiceDescriptor().getUrl(), e);
+      }
+    }
+    ;
```
Comment (youngoli) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/portable/job/ReferenceRunnerJobService.java]: Nitpick: stray semicolon?

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 2]
```diff
@@ -1389,5 +1402,39 @@ artifactId=${project.name}
...
+
+
+    /** ***********************************************************************************************/
+
+    // Method to create the PortableValidatesRunnerTask.
+    project.ext.createPortableValidatesRunnerTask = {
+      def config = it ? it as PortableValidatesRunnerConfig : new PortableValidatesRunnerConfig()
+      def name = config.name ? config.name : "validatesPortableRunner"
+      def testCategories = {
+        includeCategories 'org.apache.beam.sdk.testing.ValidatesRunner'
```
Comment (bsidhom) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: Presumably these categories will be runner-specific (with the exception of ValidatesRunner itself). Can you add the excluded categories as a configuration parameter that is specified when calling `applyJavaNature`?

Comment (bsidhom) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: Sorry, make that `createPortableValidatesRunnerTask` rather than `applyJavaNature`.

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 3]
```diff
@@ -1389,5 +1402,39 @@ artifactId=${project.name}
...
+        excludeCategories 'org.apache.beam.sdk.testing.UsesTestStream'
+      }
+      testCategories = config.testCategories ? config.testCategories : testCategories
+      project.tasks.create(name: name, type: Test) {
+        group = "Verification"
+        description = "Validates the PortableRunner with JobServer ${config.jobServerDriver}"
+        systemProperty "beamTestPipelineOptions", JsonOutput.toJson([
+          "--runner=org.apache.beam.runners.reference.testing.TestPortableRunner",
+          "--jobServerDriver=${config.jobServerDriver}",
+          config.jobServerConfig ? "--jobServerConfig=${config.jobServerConfig}" : "",
```
Comment (bsidhom) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: It might be better to construct the json list out-of-band in order to avoid the empty arg.

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 4]
```diff
@@ -1389,5 +1402,39 @@ artifactId=${project.name}
...
+        ])
+        classpath = project.configurations.validatesPortableRunner
```
Comment (bsidhom) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: Please either document the requirement of the existence of a `validatesPortableRunner` configuration or make the configuration pluggable.

File: runners/direct-java/build.gradle [hunk 5]
```diff
@@ -147,3 +156,5 @@ createJavaExamplesArchetypeValidationTask(type: 'MobileGaming',
   gcsBucket: gcsBucket,
   bqDataset: bqDataset,
   pubsubTopic: pubsubTopic)
+
+createPortableValidatesRunnerTask(name: "validatesPortableRunner", jobServerDriver: "org.apache.beam.runners.direct.portable.job.ReferenceRunnerJobServer")
```
Comment (bsidhom) [inline runners/direct-java/build.gradle]: Looks like there's no trailing newline. I guess `./gradlew spotlessApply` doesn't fix it because it's not a Java file.

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy (hunk 2 above)
Comment (angoenka) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: Runner can provide test categories in config.testCategories so they can control the appropriate test category.
Should I add only the exclude categories in the parameters?

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy (hunk 4 above)
Comment (angoenka) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: I will make it pluggable and document it.

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 6]
```diff
@@ -184,6 +185,18 @@ class BeamModulePlugin implements Plugin<Project> {
     String tag = null // Sets the image tag (optional).
   }
 
+  // A class defining the configuration for PortableValidatesRunner.
+  class PortableValidatesRunnerConfig {
```
[the remark this answers is not in the document]
Comment (angoenka) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: Done

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy (hunk 3 above)
Comment (angoenka) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: Done

File: runners/direct-java/build.gradle (hunk 5 above)
Comment (angoenka) [reply runners/direct-java/build.gradle]: done

**Comment by angoenka** [comment]
Ping for review :)

**Comment by angoenka** [comment]
cc: @tweise

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 7]
```diff
@@ -184,6 +186,20 @@ class BeamModulePlugin implements Plugin<Project> {
     String tag = null // Sets the image tag (optional).
   }
 
+  // A class defining the configuration for PortableValidatesRunner.
+  class PortableValidatesRunnerConfiguration {
```
Comment (youngoli) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: I feel like it would be better to add the defaults for this configuration object here instead of within the createPortableValidatesRunnerTask definition for clarity to people who are modifying the gradle files. That way they can see the defaults without having to read through the method, and it removes a bunch of the ?: statements in the function definition.

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 8]
```diff
@@ -1389,5 +1405,42 @@ artifactId=${project.name}
         args argsNeeded
       }
     }
+
+
+    /** ***********************************************************************************************/
+
+    // Method to create the PortableValidatesRunnerTask.
+    project.ext.createPortableValidatesRunnerTask = {
```
Comment (youngoli) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: Can it be made more clear that the parameter for this closure is supposed to be a PortableValidatesRunnerConfiguration? While reading the code it took me a while to figure out that this function does expect that class as a parameter. If there's no downside to doing so, maybe use an explicit parameter like so:

```
project.ext.createPortableValidatesRunnerTask = { PortableValidatesRunnerConfiguration it -> ... }
```

**Comment by angoenka** [comment]
Run Java PreCommit

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy (hunk 8 above)
Comment (youngoli) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: Comment sounds good to me.

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 9]
```diff
@@ -193,6 +195,30 @@ class BeamModulePlugin implements Plugin<Project> {
...
+    // Task name for validate runner case.
+    String name = 'validatesPortableRunner'
+    // Fully qualified JobServerClass name to use.
+    String jobServerDriver
+    // A string representing the jobServer Configuration.
+    String jobServerConfig
+    // Categories for tests to run.
+    Closure testCategories = {
+      includeCategories 'org.apache.beam.sdk.testing.ValidatesRunner'
+      excludeCategories 'org.apache.beam.sdk.testing.FlattenWithHeterogeneousCoders'
```
Comment (bsidhom) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: Should these default excludes apply to _all_ runners? If not, I think it makes sense to put this on each runner to specify.

Comment (angoenka) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: Yes, the exclusions here are not supported by fnexecution at the moment hence they apply to all portable validate runners.
As we support more functionality to portability layer, we will remove these exclusions.

**Comment by angoenka** [comment]
Run Java PreCommit

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy (hunk 9 above)
Comment (bsidhom) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: OK, that makes sense.

**Comment by angoenka** [comment]
Run Java PreCommit

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 10]
```diff
@@ -1394,5 +1420,31 @@ artifactId=${project.name}
...
+      ]
+      if(config.jobServerConfig){
+        beamTestPipelineOptions.add("--jobServerConfig=${config.jobServerConfig}")
+      }
+      project.tasks.create(name: name, type: Test) {
+        group = "Verification"
+        description = "Validates the PortableRunner with JobServer ${config.jobServerDriver}"
+        systemProperty "beamTestPipelineOptions", JsonOutput.toJson(beamTestPipelineOptions)
+        classpath = config.testClasspathConfiguration
+        testClassesDirs = project.files(project.project(":beam-sdks-java-core").sourceSets.test.output.classesDirs, project.project(":beam-runners-core-java").sourceSets.test.output.classesDirs)
```
Comment (bsidhom) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: Ah, I just realized that it's possible this will fail on submodules that have not already added an evaluation dependency on beam-sdks-java-core (necessary as we're accessing this source set here).

@lukecwik Do you know if there's a way to do this from within the plugin? Is that even recommended if possible? Can we instead hard fail with a message if the evaluation dependency has not been specified?

Comment (bsidhom) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: Another question about the structure here: Should we add this within `applyJavaNature`? The `ValidatesRunner` tests only work for Java-based runners anyway (and only the only runners that exist are written in Java).

Comment (bsidhom) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: Scratch that last comment. I misread the surrounding context.

Comment (angoenka) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: I suppose the gradle task will fail verbosely if the down stream project does not have dependency on :beam-sdks-java-core 
The best I can think of at the moment is just document expectation for :beam-sdks-java-core dependency.

Comment (bsidhom) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: OK, that sounds good for now.

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 11]
```diff
@@ -1425,6 +1425,7 @@ artifactId=${project.name}
     /** ***********************************************************************************************/
 
     // Method to create the PortableValidatesRunnerTask.
+    // Generated task expect :beam-sdks-java-core and :beam-runners-core-java to be in the class path.
```
Comment (bsidhom) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: Reword this to make it clear this is a gradle evaluation dependency (rather than a java classpath issue).

Specifically, users must make sure the following is in the consuming module's `build.gradle` file:

```
evaluationDependsOn(":beam-sdks-java-core")
evaluationDependsOn(":beam-runners-core-java")
```

**Comment by lukecwik** [comment]
Taking a look.

**Comment by tweise** [comment]
It would be nice to squash merge or tag the main commits with the JIRA number.

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 12]
```diff
@@ -1392,5 +1418,32 @@ artifactId=${project.name}
...
+    // Generated task expect :beam-sdks-java-core and :beam-runners-core-java in the evaluation dependency. More specifically the task expects evaluationDependsOn(":beam-sdks-java-core") and evaluationDependsOn(":beam-runners-core-java") in the calling build.gradle file.
+    // The method takes PortableValidatesRunnerConfiguration as parameter.
+    project.ext.createPortableValidatesRunnerTask = {
+      def config = it ? it as PortableValidatesRunnerConfiguration : new PortableValidatesRunnerConfiguration()
+      def name = config.name
+      def beamTestPipelineOptions = [
+        "--runner=org.apache.beam.runners.reference.testing.TestPortableRunner",
+        "--jobServerDriver=${config.jobServerDriver}",
+      ]
+      if(config.jobServerConfig){
```
Comment (lukecwik) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: nit: spacing around ){

**Comment by angoenka** [comment]
Thanks Thomas.
Squashed the commits.

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy (hunk 10 above)
Comment (lukecwik) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: you should be able to add:
```
project.evaluationDependsOn(":beam-sdks-java-core")
project.evaluationDependsOn(":beam-runners-core-java")
```

Don't forget to update the comment as well.

File: runners/direct-java/build.gradle [hunk 13]
```diff
@@ -65,6 +66,7 @@ dependencies {
   shadow project(path: ":beam-sdks-java-core", configuration: "shadow")
   shadow library.java.joda_time
   shadow library.java.slf4j_api
+  shadow library.java.slf4j_simple
```
Comment (lukecwik) [inline runners/direct-java/build.gradle]: Why do you want to add slf4j_simple here?

You typically only add SLF4J logging backends as test / application dependencies. (e..g add it to the validatesPortableRunner configuration)

File: runners/flink/job-server/build.gradle [hunk 14]
```diff
@@ -40,15 +40,15 @@ apply plugin: "application"
 mainClassName = "org.apache.beam.runners.flink.FlinkJobServerDriver"
```
Comment (lukecwik) [inline runners/flink/job-server/build.gradle]: we should be able to drop the evaluationDependsOn above because it was moved into the task.

File: runners/direct-java/build.gradle (hunk 13 above)
Comment (angoenka) [reply runners/direct-java/build.gradle]: That was left from some debugging. Cleaning it up.

File: runners/flink/job-server/build.gradle (hunk 14 above)
Comment (angoenka) [reply runners/flink/job-server/build.gradle]: Makes sense.

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy (hunk 12 above)
Comment (angoenka) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: Done

**Comment by angoenka** [comment]
Ping.
All comments addressed and the tests are green :)
